src/cljam/io/pileup.clj

```diff
@@ -17,7 +17,8 @@
                        ^boolean end?
                        insertion ;; String?
                        deletion ;; int?
-                       qname]) ;; String?
+                       qname ;; String?
+                       alignment]) ;; SAMAlignment?
```

Comment:
Thank you for sharing.
The `PileupBase` change is the most effective change in this PR.
`assoc` an alignment in `resolve-base` takes a long time in my profiling.
That's why I've proposed adding `alignment` field to `PileupBase`.

I measured the performance of master with only the `PileupBase` change.

- sample: WES SRP006438.sorted.bam
  
    - many alignments chr1:152210000-152214999

    |branch|time|
    |----|----|
    |master|time: 7.490895 sec, sd: 46.194343 ms|
    |master w/ PileupBase chage|time: 6.910773 sec, sd: 108.074591 ms|

    - gene: NRAS chr1:114703000-114723000
    
    |branch|time|
    |----|----|
    |master|time: 1.706070 sec, sd: 19.496387 µs|
    |master w/ PileupBase chage|time: 1.456302 sec, sd: 42.453708 µs|

    - gene: EGFR chr7:55015000-55215000
    
    |branch|time|
    |----|----|
    |master|time: 4.215258 sec, sd: 47.407546 µs|
    |master w/ PileupBase chage |time: 3.629386 sec, sd: 284.862544 µs|

Reply:
Thank you for elaborating on the situation!
I think I underestimated the cost of assoc'ing the optional field with a record. To double check your benchmark result, I carried out an additional examination with several configurations.

The following is the examination result:

![Direct liniking disabled](https://user-images.githubusercontent.com/27441/58617081-283c8300-82fa-11e9-88a3-f30f34aaf768.png)

![Direct linking enabled](https://user-images.githubusercontent.com/27441/58617052-18bd3a00-82fa-11e9-9e9e-add82685f34d.png)

The x-axis in these charts is representing the JVM option passed to specify the maximum heap size. The series `master`, `feature/improve-pileup` show the result from the corresponding branches. The series named `optional :alignment` shows the result from a branch that adds the change to the `feature/improve-pileup` branch making the `alignment` field optional. The first chart was conducted with direct linking disabled while the second one with direct linking enabled. For more detailed data, see [here](https://docs.google.com/spreadsheets/d/1mPokNb7cEqghMywao0uNmgjzM56m7gVRMj96PwnXSm8/edit?usp=sharing).

I'm not sure how accurate the detailed figures are, but I think we can see their rough trends. 
As a result, it seems the `optional :alignment` only contributes to very limited performance gain in contexts with relatively high memory pressure, although it can also be said that it does contribute to as much portion of the performance gain as the other stuff does in contexts with sufficient heap assignment.

That being said, pileup performance improvement by 10-20% in total is no trivial and is fairly attractive, so now I agree it's worth making the `alignment` a required field. 😃